keps/sig-node/1999-checkpoint-restore/README.md

```diff
@@ -0,0 +1,277 @@
+   transfer, restore): https://criu.org/Podman
+
+   The reason for mentioning Podman in this list is that adding
+   checkpoint and restore to Podman can be seen as a preparation for
+   this KEP. Especially as Podman uses CNI for network configuration
+   it required changes to CRIU and runc to allow restoring into a
+   previously set up network namespace instead of letting CRIU handle
+   the network namespace restore:
+
+   * https://github.com/opencontainers/runc/pull/1849
+
+   This work lead to CRIU's support to allow restoring processes
+   out of and into PID namespaces:
+
+   * https://github.com/checkpoint-restore/criu/pull/1056
+   * https://github.com/opencontainers/runc/pull/2525
+
+   As well as the runc support to allow restoring into all kinds of
+   namespaces: https://github.com/opencontainers/runc/pull/2583
+
+   The support to allow restoring into all kinds of namespaces is
+   especially important when checkpointing and restoring containers
+   out of and into pods.
+
+The motivation is to get closer to container migration by taking the first
+step and providing support for simple checkpoint and restore. The motivation
+is existing checkpoint/restore/migration support in other container
+environments showing that it is useful and production ready.
+
+### Goals
+
+The goal of this KEP is to introduce *checkpoint* and *restore* to the API.
+
+### Non-Goals
+
+Out of scope of this KEP are high level discussions about how to implement
+container migration. This is only about low level primitives to add
+*checkpoint* and *restore* to the API.
+
+
+## Proposal
+
+```patch
+--- a/api.proto
++++ b/api.proto
+@@ -87,6 +87,11 @@ service RuntimeService {
+     // Once it returns error, new container log file MUST NOT be created.
+     rpc ReopenContainerLog(ReopenContainerLogRequest) returns (ReopenContainerLogResponse) {}
+ 
++    // CheckpointContainer checkpoints a container
++    rpc CheckpointContainer(CheckpointContainerRequest) returns (CheckpointContainerResponse) {}
++    // RestoreContainer restores a container
++    rpc RestoreContainer(RestoreContainerRequest) returns (RestoreContainerResponse) {}
++
+     // ExecSync runs a command in a container synchronously.
+     rpc ExecSync(ExecSyncRequest) returns (ExecSyncResponse) {}
+     // Exec prepares a streaming endpoint to execute a command in the container.
+@@ -1308,3 +1313,23 @@ message ReopenContainerLogRequest {
+ 
+ message ReopenContainerLogResponse{
+ }
++
++message CheckpointContainerRequest {
++    // ID of the container to be checkpointed
++    string container_id = 1;
++}
++
++message CheckpointContainerResponse{
++}
++
++message RestoreContainerRequest {
++    // ID of the container to be restored
++    string container_id = 1;
++    // ID of the sandbox into which the container should be restored
++    // If this is not specified the container will be restored
++    // in the sandbox it was checkpointed from.
++    string pod_sandbox_id = 2;
```

Comment:
The runtimes I've investigated (Docker, containerd and a tiny bit of cri-o) seem to require to create a container before restoring a checkpoint. If a checkpoint location can be specified, I would remove this parameter, as the sandbox can be changed by creating a container in a different sandbox.
It's probably also more in line with the way kubernetes works, as it's not possible to move containers between pods right now, and would probably be very hard to integrate into the API.
Finally this would probably lead to some duplicated code in each runtime implementation (modifying the sandbox the container belongs to) that could also be part of the client side.

On the other hand, creating a new container to change the sandbox could lead to issues with the container image, which is probably where this approach is coming from: If the image is using a non-specific tag like `:latest`, creating a new container could prevent the checkpoint from being restored successfully, if the image changed significantly.

Reply:
> The runtimes I've investigated (Docker, containerd and a tiny bit of cri-o) seem to require to create a container before restoring a checkpoint.

The whole concept of different implementations that a container needs to be created first before restoring always felt wrong. It seems like a pretty obvious step. If the container does not exist during restore, just let the runtime create it.

The way I have done it in Podman is to create the container on restore if it does not exist. That way it is easier for the user. To require a manual step of first creating a container does not really seem useful.

> If a checkpoint location can be specified, I would remove this parameter, as the sandbox can be changed by creating a container in a different sandbox.

As mentioned above. Why requiring an additional step of creating a container which can easily done by the runtime.

> Finally this would probably lead to some duplicated code in each runtime implementation (modifying the sandbox the container belongs to) that could also be part of the client side.

It can still be part of the client side, but I would not remove it because moving to another sandbox is easy to do during restore. runc and CRIU can handle it in their latest version.

> On the other hand, creating a new container to change the sandbox could lead to issues with the container image, which is probably where this approach is coming from: If the image is using a non-specific tag like :latest, creating a new container could prevent the checkpoint from being restored successfully, if the image changed significantly.

Again I need to mention my Podman work. Podman checkpoints include everything. The CRIU checkpoint, information about CNI, config.json and the file system diff. Using `podman container checkpoint --export=export.tar.gz` gives you one single tarball which can be moved to any other machine to restore the container. All necessary information is extracted to restore a container with one single command `podman container restore --import=export.tar.gz`. Requiring the user to first manually a container is not necessary from the implementation side. To restore a container into a non-existing image, Podman will `pull` the image using its ID and not using the name and the tag. As long as the image is still in the registry the restore will work. All changed files on the original container's file system will also be updated. See: https://criu.org/Podman

I see no problems, however, to add fields to the API proposal you need for your work, but I would not like to remove `pod_sandbox_id`.

Reply:
> Podman checkpoints include everything. The CRIU checkpoint, information about CNI, config.json and the file system diff. Using podman container checkpoint --export=export.tar.gz gives you one single tarball which can be moved to any other machine to restore the container. All necessary information is extracted to restore a container with one single command podman container restore --import=export.tar.gz. Requiring the user to first manually a container is not necessary from the implementation side.

In the context of Kubernetes this would lead to the question where the spec/"truth" comes from. As far as I understand, there is a very strict "API is the source of truth" rule, which is only rarely broken. As it would probably be very hard to implement actually moving Pods between hosts through the API, my approach was to allow cloning Pods instead. A Pod can contain a reference to a source Pod that is supposed to be cloned. If the source Pod is deleted in the process, it's essentially a migration. The cloning only includes the process within the containers though. The Pod still needs to contain container specs as usual and probably should allow to modify anything that doesn't break checkpoint/restore.
The same applies to other use-cases of checkpoint-restore as well: The container specification should be defined in the API, not by a checkpoint that might look completely different and lead to state drift.